### DISCUSSION TITLE
Issue with optimistic UI being overwritten by the return of an old mutation.

### DISCUSSION THREAD

**Discussion by taylorcantwell** [urql-graphql/urql#2169] [Q&A] [answered]
Hi guys,

I'm trying to achieve something here and having a little trouble, and hoping someone could point me in the right direction or correct me if I'm going about this the wrong way.

What I'm trying to achieve is a button that you can freely activate that sends out a mutation and also optimistically updates the UI. However, when you spam it sending out multiple mutations, the most recent optimistic UI (the last expected result) is overwritten by the cache update when a mutations result is returned. So, for example, with a button that toggles something on and off you would expect on --> off--> on--> off. Instead it's unpredictable.

A solution I thought of was checking to see if the mutation is pending and if it is simply return the current cache that was set by the optimistic update - however, it doesn't appear you can do that? I'm trying to avoid limiting the frequency you can press the button and would like to keep the state in the cache.

The code below is the cache updates I'm trying to achieve:

```
  likeTweet: (result, { tweetId }, cache, info) => {
            cache.updateQuery({ query: TweetsDocument }, (data) => {
              const indexOfLikedTweet = data.tweets.findIndex(
                (tweet) => tweet.id === tweetId
              );

              data.tweets[indexOfLikedTweet].likeStatus = true;
  
              return { tweets: data.tweets };
            });
          },
```

```
unlikeTweet: (result, { tweetId }, cache, info) => {
            cache.updateQuery({ query: TweetsDocument }, (data) => {
              const indexOfLikedTweet = data.tweets.findIndex(
                (tweet) => tweet.id === tweetId
              );

              data.tweets[indexOfLikedTweet].likeStatus = false;
  
              return { tweets: data.tweets };
            });
          },
```

**Comment by kitten** [comment, marked as answer]
Have you got more code here? You're showing updating a list, however, if I'm not mistaken and you only have a single flag that indicates the "liked" status then you'd only need to update the liked item.
So, youd only update the tweet that's changing 

That said, I can also not see your "optimistic" config which would actually be responsible for making the optimistic change. This feature is also built to be handling "chaotic" state updates like these where many updates happen at the same time 

That said, since the mutations will then run concurrently, it's pretty important to either make sure that they can never conflict, or to make the mutually exclusive or serial with a custom exchange 
(That said, the UI can also just block while the mutation is in progress)

However, with the `optimistic` config you're solving the first problem; which is that before the mutations executes and is sent, the state will be consistent 

https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#optimistic-updates

**Reply by taylorcantwell** [reply]
Sorry I should have provided some more information. There are two mutations: one to like and another like which and are triggered depending on the like status.


Here is the optimistic logic:
```
    cacheExchange({
      optimistic: {
        createTweet: createTweetCacheUpdate,
        likeTweet: (variables, cache, info) => {
          return true;
        },
        unlikeTweet: (variables, cache, info) => {
          return true;
        },
      },
```

I'm simply returning true for now if the mutation is successful.

**Reply by mstrchng** [reply]
@taylorcantwell hello, did you manage to solve this?

**Reply by taylorcantwell** [reply]
@mstrchng Sorry for the late response. I haven't touched this since posting this. I might swing back to it at some point though. Did you end up working it out?

**Reply by PatrykWalach** [reply]
I think, I'm experiencing the same issue.
I expect `Media.isFavourite` to be true when the Media is in the list returned by the mutation.

After I click twice the optimistic UI works as intended.
But when the requests come back I get a flicker.

When the first mutation returns I would expect the response to be overridden by the second optimistic mutation.

```jsx
({
	updates: {
		Mutation: {
			ToggleFavourite(result, args, cache) {
				if (!result.ToggleFavourite?.media?.nodes) {
					return
				}

				const isFavourite = result.ToggleFavourite.media.nodes.some((media) => media?.id === Number(args.mediaId))

				cache.writeFragment(gql(/* GraphQL */`
				fragment _ on Media {
					id
					isFavourite
				}
				`), {
					id: args.mediaId,
					isFavourite,
				})
			}
		}
	},
	optimistic: {
		ToggleFavourite(args, cache) {
			const isFavourite = cache.resolve({
				__typename: 'Media',
				id: args.mediaId
			}, 'isFavourite')

			return {
				__typename: 'Favourites',
				media: {
					__typename: 'MediaConnection',
					nodes: !isFavourite ? [{
						id: Number(args.mediaId),
						__typename: 'Media',
					}] : []
				}
			}
		},
	}
})
```

EDIT: I fixed this by passing [extra variables](https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#variables-for-optimistic-updates)
```diff
+ const isFavourite = info.variables.isFavourite
- const isFavourite = cache.resolve(({
- __typename: 'Media',
- id: args.mediaId
- }), 'isFavourite')
```
